packages/core/src/operator/operator-runtime/listenForChallengesCallback.ts

```diff
@@ -0,0 +1,114 @@
+                operatorState.cachedLogger(`${error.message}.`);
+            });
+    }
+
+    operatorState.cachedLogger(`Received new challenge with number: ${challengeNumber}.`);
+    operatorState.cachedLogger(`Processing challenge...`);
+
+    // Add a delay of 1 -300 seconds to the new challenge process so not all operators request the subgraph at the same time
+    const delay = Math.floor(Math.random() * 301);
+    await new Promise((resolve) => {
+        setTimeout(resolve, delay * 1000);
+    })
+
+    try {
+        const graphStatus = await getSubgraphHealthStatus();
+        if (graphStatus.healthy) {
+            const { sentryWalletMap, sentryKeysMap, nodeLicenseIds, mappedPools, refereeConfig } =
+                await loadOperatorKeysFromGraph(operatorState.operatorAddress, challengeNumber - 1n);
+
+            await processNewChallenge(challengeNumber, challenge, nodeLicenseIds, sentryKeysMap, sentryWalletMap, mappedPools, refereeConfig);
+            // check the previous challenge, that should be closed now
+            if (challengeNumber > BigInt(1)) {
+                await processClosedChallenges(challengeNumber - BigInt(1), nodeLicenseIds, sentryKeysMap, sentryWalletMap);
+            }
+
+
+        } else {
+            operatorState.cachedLogger(`Revert to RPC call instead of using subgraph. Subgraph status error: ${graphStatus.error}`)
+
+            const { sentryKeysMap, nodeLicenseIds } = await loadOperatorKeysFromRPC(operatorState.operatorAddress);
+
+            await processNewChallenge(challengeNumber, challenge, nodeLicenseIds, sentryKeysMap);
+            // check the previous challenge, that should be closed now
+            if (challengeNumber > BigInt(1)) {
+                await processClosedChallenges(challengeNumber - BigInt(1), nodeLicenseIds, sentryKeysMap);
+            }
+        }
+    } catch (error: any) {
+        operatorState.cachedLogger(`Error processing new challenge in listener callback: - ${error && error.message ? error.message : error}`);
+    }
+}
+
+// Helper function to request the assertion from the public node's CDN
+async function getPublicNodeFromBucket(confirmHash: string) {
+    const url = `https://sentry-public-node.xai.games/assertions/${confirmHash.toLowerCase()}.json`;
+    const response = await axios.get(url);
+
+    if (response.status === 200) {
+        return response.data;
+    } else {
+        throw new Error("Invalid response status " + response.status);
+    }
+}
+
+/**
+ * Compare a challenge with an assertion posted to the public CDN by the public Xai node.
+ * @param {Challenge} challenge - The challenge from the Referee contract.
+ * @returns {Promise<() => Promise<{ publicNodeBucket: PublicNodeBucketInformation, error?: string }>>} Returns the assertion data from the CDN or an error on miss match.
+ */
+async function compareWithCDN(challenge: Challenge): Promise<{ publicNodeBucket: PublicNodeBucketInformation, error?: string }> {
```

Comment:
Should this be in a separate file in utils?

Reply:
Its only every used for the listenForChallenge callback that why I would have it in the same file, but I leave that call up to you.